File: JavaScript/interview_preparation_kit/problem_2_counting_valleys.js
function countingValleys(steps, path) {
  let valleys = 0;
  let level = 0;
  for (let i = 0; i < steps; i++) {
    if (path[i] === "U") {
      level++;
      if (level === 0) valleys++;
    } else {
      level--;
    }
  }
  return valleys;
}

const steps = 8;
const path = "UDDDUDUU";
console.time("Execution Time");
const result = countingValleys(steps, path);
console.timeEnd("Execution Time");
console.log(result);

File: python/interview_preparation_kit/problem_2_counting_valleys.py
import time


def countingValleys(steps, path):
    level = 0
    valleys = 0
    for i in range(steps):
        if path[i] == "U":
            level += 1
            if level == 0:
                valleys += 1
        else:
            level -= 1
    return valleys


if __name__ == "__main__":
    steps = 8
    string = "UDDDUDUU"
    start_time = time.time()
    result = countingValleys(steps, string)
    print(result)
    elapsed_time = time.time() - start_time
    print(f"--- {elapsed_time:.4f} seconds ---")
